op/charm.py

```diff
@@ -37,40 +37,52 @@ class LeaderSettingsChangedEvent(HookEvent):
 
 
 class RelationEvent(HookEvent):
-    def __init__(self, handle, relation):
+    def __init__(self, handle, relation, app=None, unit=None):
         super().__init__(handle)
+
+        if unit and unit.app != app:
+            raise RuntimeError(f'cannot create RelationEvent with application {app} and unit {unit}')
+
         self.relation = relation
+        self.app = app
+        self.unit = unit
 
     def snapshot(self):
-        return {
+        snapshot = {
             'relation_name': self.relation.name,
             'relation_id': self.relation.id,
         }
 
-    def restore(self, snapshot):
-        self.relation = self.framework.model.get_relation(snapshot['relation_name'], snapshot['relation_id'])
+        if self.app:
+            snapshot['app_name'] = self.app.name
 
-class RelationUnitEvent(RelationEvent):
-    def __init__(self, handle, relation, unit):
-        super().__init__(handle, relation)
-        self.unit = unit
+        if self.unit:
+            snapshot['unit_name'] = self.unit.name
```

Comment:
One more trivial nitpick: this function will likely look better with all of the empty lines dropped.